### PR TITLE
Disallow duplicate named exports

### PR DESCRIPTION
**Pull request by kaicataldo**
I think I got all the cases for named exports - let me know if I missed one! Also wasn't sure about the error message wording, but was hoping that by including the name of the offending duplicate export that debugging could be made easier for the end user. Finally, wasn't sure where to put the tests - happy to move them!

### CODE REVIEW

File: test/fixtures/flow/type-exports/interface/actual.js [hunk 1]
```diff
@@ -1,2 +1,2 @@
 export interface foo { p: number };
-export interface foo<T> { p: T };
+export interface bar<T> { p: T };
```
Comment (kaicataldo) [inline test/fixtures/flow/type-exports/interface/actual.js]: I'm pretty new to Flow - let me know if this isn't the correct fix for this test

**Comment by kaicataldo** [comment]
Looks like it's failing because of test files in Babel plugins that contain multiple exports in the same file. How is that usually handled? I'm happy to make the necessary changes in the Babel repo too.

**Comment by danez** [comment]
Good job. That looks really good. The flow test looks good. 👍 

The tests of babel now fail, but this is okay for now, but we would need to correct them at some point, before we release this.

**Comment by kaicataldo** [comment]
Cool! Should I go ahead and make a PR to the main babel repo?

**Comment by danez** [comment]
Yes that would be nice. Thanks.

Oh for some reason I did not see the other comment of you. So there is no real workflow yet on how to do the cross-repo stuff. I described in this PR how to link babylon in babel. I should really but that in the Contributing file.

https://github.com/babel/babel/pull/3700#issuecomment-244622489

**Comment by kaicataldo** [comment]
Thanks for all the info on this! One more question - do you think I should split the changes into batches? Splitting each one of those files into multiple files is going to be a pretty massive change.

**Comment by danez** [comment]
Ah sorry, for not answering. As long as the changes belong together I think they should go into the same PR imho. And I think this would account for this case.

**Comment by kaicataldo** [comment]
Finally started in on this, though might take me a few days. Doing it in chunks because there are a lot of changes.

**Comment by kaicataldo** [comment]
Finally got around to fixing those tests! PR open here: https://github.com/babel/babel/pull/4538

I mention it in the issue as well, but will also make a 2nd PR that removes the [initial band-aid fix](https://github.com/babel/babel/pull/3518) that was merged into Babel itself and to rename the test files so they follow the convention used in the currently open PR. I plan on getting to this tomorrow!

File: src/parser/statement.js [hunk 2]
```diff
@@ -844,7 +844,7 @@ pp.parseExport = function (node) {
     }
     node.declaration = expr;
     if (needsSemi) this.semicolon();
-    this.checkExport(node);
+    this.checkExport(node, true, true);
```
Comment (hzoo) [inline src/parser/statement.js]: just a thought: might be nice if this method wasn't taking in true, true (an object, named, etc) but I know we do that everywhere else ^ `this.parseFunction(expr, true, false, false, true);` 😄

Comment (kaicataldo) [inline src/parser/statement.js]: Yeah, was trying to follow the style I saw elsewhere, but agree. Would you like me to change that to an object?

File: test/fixtures/flow/type-exports/interface/actual.js (hunk 1 above)
Comment (danez) [inline test/fixtures/flow/type-exports/interface/actual.js]: This is correct

File: src/parser/statement.js (hunk 2 above)
Comment (danez) [inline src/parser/statement.js]: @hzoo: Do you mean `checkExport(node: Node, options: { named: boolean, default: boolean })` or what do you mean with object? Maybe also  `checkExport(node: Node, type: "named" | "default" | null)`?

Comment (hzoo) [inline src/parser/statement.js]: yeah to make it more readable - dono if it was done before for perf?

Comment (kaicataldo) [inline src/parser/statement.js]: I assumed it was done this way for performance reasons

**Comment by kaicataldo** [comment]
Opened one more PR that reverts the commit I made that throws a warning in babel itself since we have moved that into babylon. I also updated the test names so that they match the convention I used in my more recent PR that was merged today. Let me know if you need anything else!

**Comment by dlmr** [comment]
Seems like this change causes Babylon to throw for one of my files that worked fine before. 

The file in questions looks like this:
```javascript
export toString from './toString';
```

```bash
`toString` has already been exported. Exported identifiers must be unique. (1:7)
> 1 | export toString from './toString';
    |        ^
  2 |
```
Is this an intended change, is not `toString` allowed any more?

Seems like all keys on `Object.prototype` causes this error.

File: src/parser/statement.js [hunk 3]
```diff
@@ -913,6 +937,20 @@ pp.checkExport = function (node) {
   }
 };
 
+pp.checkDuplicateExports = function(node, name, isDefault) {
+  if (this.state.exportedIdentifiers[name]) {
```
Comment (dlmr) [inline src/parser/statement.js]: I think we need to manage `Object.prototype` here.

Something along the lines of this maybe?

```js
if (Object.getOwnPropertyNames(this.state.exportedIdentifiers).indexOf(name) !== -1) {
```

Comment (danez) [inline src/parser/statement.js]: You are right, I wrap up a fix

Comment (hzoo) [inline src/parser/statement.js]: Ah interesting

cc @kaicataldo

Comment (kaicataldo) [inline src/parser/statement.js]: Oh wow, that makes sense. @danez Let me know if you need me work on a fix - happy to do it.

Comment (hzoo) [inline src/parser/statement.js]: Same issue in https://github.com/jscs-dev/node-jscs/pull/1204 😄

**Pull request merged**